applications/RomApplication/custom_utilities/rom_residuals_utility.h

```diff
@@ -138,17 +140,183 @@ namespace Kratos
+            const auto el_begin = mpModelPart.ElementsBegin();
+            const auto cond_begin = mpModelPart.ConditionsBegin();
+
+            //contributions to the system
+            Matrix LHS_Contribution = ZeroMatrix(0, 0);
+            Vector RHS_Contribution = ZeroVector(0);
+
+            //vector containing the localization in the system of the different terms
+            Element::EquationIdVectorType EquationId;
+            Matrix MatrixResiduals( (nelements + nconditions), mPetrovGalerkinRomDofs); // Matrix of reduced residuals.
+            Matrix PsiElemental;
+            #pragma omp parallel firstprivate(nelements, nconditions, LHS_Contribution, RHS_Contribution, EquationId, PsiElemental, el_begin, cond_begin)
+            {
+                #pragma omp for nowait
+                for (int k = 0; k < nelements; k++){
+                    auto it_el = el_begin + k;
+                    //detect if the element is active or not. If the user did not make any choice the element is active by default
+                    bool element_is_active = true;
+                    if ((it_el)->IsDefined(ACTIVE))
+                        element_is_active = (it_el)->Is(ACTIVE);
+                    if (element_is_active){
+                        //calculate elemental contribution
+                        mpScheme->CalculateSystemContributions(*it_el, LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);
+                        Element::DofsVectorType dofs;
+                        it_el->GetDofList(dofs, CurrentProcessInfo);
+                        //assemble the elemental contribution - here is where the ROM acts
+                        //compute the elemental reduction matrix PhiElemental
+                        const auto& geom = it_el->GetGeometry();
+                        if(PsiElemental.size1() != dofs.size() || PsiElemental.size2() != mPetrovGalerkinRomDofs)
+                            PsiElemental.resize(dofs.size(), mPetrovGalerkinRomDofs,false);
+                        RomAuxiliaryUtilities::GetPsiElemental(PsiElemental, dofs, geom, MapPhi);
+                        noalias(row(MatrixResiduals, k)) = prod(trans(PsiElemental), RHS_Contribution); // The size of the residual will vary only when using more ROM modes, one row per condition
+                    }
+
+                }
+
+                #pragma omp for nowait
+                for (int k = 0; k < nconditions;  k++){
+                    ModelPart::ConditionsContainerType::iterator it = cond_begin + k;
+                    //detect if the condition is active or not. If the user did not make any choice the condition is active by default
+                    bool condition_is_active = true;
+                    if ((it)->IsDefined(ACTIVE))
+                        condition_is_active = (it)->Is(ACTIVE);
+                    if (condition_is_active){
+                        Condition::DofsVectorType dofs;
```

Comment:
Same comment. If allocated outside you can use the same `dofs` variable you use for the elements.

Reply:
Done